### PR TITLE
perf: Add option of stereo angle when building telescope detector

### PR DESCRIPTION
**Pull request by XiaocongAi**
This PR add options for stereo angles of each telescope plane when building the telescope detector, which is helpful in case that the telescope planes are strip detectors with stereo angles.

### CODE REVIEW

File: Examples/Python/tests/test_detectors.py [hunk 1]
```diff
@@ -44,6 +44,7 @@ def test_telescope_geometry():
     detector, geo, contextDecorators = acts.examples.TelescopeDetector.create(
         bounds=[100, 100],
         positions=[10 * i for i in range(n_surfaces)],
+        stereos=[0 for i in range(n_surfaces)],
```
Comment (paulgessinger) [inline Examples/Python/tests/test_detectors.py:47]:
```suggestion
        stereos=[0] * n_surfaces,
```

File: Examples/Scripts/Python/truth_tracking_telescope.py [hunk 2]
```diff
@@ -11,19 +11,19 @@
 
 if "__main__" == __name__:
     detector, trackingGeometry, decorators = acts.examples.TelescopeDetector.create(
-        bounds=[200, 200], positions=[30, 60, 90, 120, 150, 180, 210, 240, 270]
+        bounds=[200, 200],
+        positions=[30, 60, 90, 120, 150, 180, 210, 240, 270],
+        stereos=[0, 0, 0, 0, 0, 0, 0, 0, 0],
```
Comment (paulgessinger) [inline Examples/Scripts/Python/truth_tracking_telescope.py:16]: Maybe 
```suggestion
        stereos=[0]*9,
```

**Review by paulgessinger — COMMENTED** [review]
Looks good

**Comment by paulgessinger** [comment]
I applied my own suggestions, now my review is not eligible anymore. @felix-russo or @AJPfleger could you have a quick look at this?

File: Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp [hunk 3]
```diff
@@ -34,7 +34,8 @@ enum class TelescopeSurfaceType {
 ///
 /// @param gctx is the detector element dependent geometry context
 /// @param detectorStore is the store for the detector element
-/// @param positions is the offset w of different layers in the longitudinal
+/// @param positions is the offset of different layers in the longitudinal
```
Comment (felix-russo) [inline Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp:37]:
```suggestion
/// @param positions is the offset of different layers in the longitudinal direction
```

File: Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp [hunk 4]
```diff
@@ -34,7 +34,8 @@ enum class TelescopeSurfaceType {
...
+/// @param stereos is the stereo angle of different layers around the longitudinal
```
Comment (felix-russo) [inline Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp:38]:
```suggestion
/// @param stereos is the stereo angle (i.e., ...) of different layers around the longitudinal direction
```

Comment (felix-russo) [reply Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp:38]: Could you maybe give a short explanation of what a stereo angle is? I guess it is something like the polar angle phi?

File: Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp [hunk 5]
```diff
@@ -46,9 +47,9 @@ enum class TelescopeSurfaceType {
 std::unique_ptr<const Acts::TrackingGeometry> buildDetector(
     const typename TelescopeDetectorElement::ContextType& gctx,
     std::vector<std::shared_ptr<TelescopeDetectorElement>>& detectorStore,
-    const std::vector<double>& positions, const std::array<double, 2>& offsets,
-    const std::array<double, 2>& bounds, double thickness,
-    TelescopeSurfaceType surfaceType,
+    const std::vector<double>& positions, const std::vector<double>& stereos,
```
Comment (felix-russo) [inline Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp:50]: Should we consider renaming the parameter `stereo` -> `stereoAngles`?

File: Examples/Detectors/TelescopeDetector/src/BuildTelescopeDetector.cpp [hunk 6]
```diff
@@ -80,8 +80,14 @@ ActsExamples::Telescope::buildDetector(
   for (unsigned int i = 0; i < nLayers; ++i) {
     // The translation without rotation yet
     Acts::Translation3 trans(offsets[0], offsets[1], positions[i]);
-    // The transform
+    // The transform (the center defined by trans will be rotated as well with
+    // the following definition of trafo)
```
Comment (felix-russo) [inline Examples/Detectors/TelescopeDetector/src/BuildTelescopeDetector.cpp:84]:
```suggestion
    // The entire transformation (the coordinate system, whose center is defined by trans, will be rotated as well)
```

File: Examples/Detectors/TelescopeDetector/src/TelescopeDetector.cpp [hunk 7]
```diff
@@ -35,16 +35,23 @@ auto ActsExamples::Telescope::TelescopeDetector::finalize(
         "The minR should be smaller than the maxR for disc surface bounds.");
   }
 
+  if (cfg.positions.size() != cfg.stereos.size()) {
+    throw std::invalid_argument(
+        "The size of provided positions must be the same as the size of "
+        "provided stereos.");
```
Comment (felix-russo) [inline Examples/Detectors/TelescopeDetector/src/TelescopeDetector.cpp:41]:
```suggestion
        "The number of provided positions must match the number of "
        "provided stereo angles.");
```

File: Examples/Detectors/TelescopeDetector/src/BuildTelescopeDetector.cpp [hunk 8]
```diff
@@ -80,8 +80,14 @@ ActsExamples::Telescope::buildDetector(
...
     Acts::Transform3 trafo(rotation * trans);
+
+    // twist around local z axis by stereo angle
```
Comment (felix-russo) [inline Examples/Detectors/TelescopeDetector/src/BuildTelescopeDetector.cpp:87]:
```suggestion
    // rotate around local z axis by stereo angle
```

File: Examples/Run/Common/src/TelescopeDetectorWithOptions.cpp [hunk 9]
```diff
@@ -24,6 +24,11 @@ void TelescopeDetectorWithOptions::addOptions(
       value<VariableReals>()->default_value({{0, 30, 60, 120, 150, 180}}),
       "Telescope detector Input: the layers positions in the longidutinal "
       "direction in mm");
+  opt("geo-tele-stereos",
+      value<VariableReals>()->default_value({{0, 0, 0, 0, 0, 0}}),
+      "Telescope detector Input: the layers stereo angle around the "
+      "longidutinal "
```
Comment (felix-russo) [inline Examples/Run/Common/src/TelescopeDetectorWithOptions.cpp:30]:
```suggestion
      "longitudinal "
```

**Review by felix-russo — COMMENTED** [review]
Looks good, just a few (probably overly pedantic) remarks from my side

File: Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp (hunk 4 above)
Comment (paulgessinger) [reply Examples/Detectors/TelescopeDetector/include/ActsExamples/TelescopeDetector/BuildTelescopeDetector.hpp:38]: To me the descriptions seems relative clear: rotation around the longitudinal direction. I'll try to clarify it a bit more, anyway.

**Comment by paulgessinger** [comment]
@felix-russo can you check back if you're happy with the changes?

**Review by felix-russo — APPROVED** [review]

**Comment by acts-project-service** [comment]
## :white_check_mark: Athena integration test results
### :white_check_mark: All tests successful

|status|job|
|------|---|
|:green_circle: | [run_unit_tests](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537656) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsWorkflow.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537655) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateTracks.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537654) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsCoreSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537653) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateActsSpacePoints.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537652) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537651) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateOrthogonalSeeds.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537650) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsValidateClusters.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537649) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsPersistifyEDM.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537648) |
|:green_circle: | [run_ci_tests: ../athena/AtlasTest/CITest/test/ActsKfRefitting.sh](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537647) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsExtrapolationAlgTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537646) |
|:green_circle: | [run_ci_tests: python3 ../athena/Tracking/Acts/ActsGeometry/test/ActsITkTest.py](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537645) |
|:green_circle: | [run_workflow_tests_run4_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537644) |
|:green_circle: | [run_workflow_tests_run2_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537643) |
|:green_circle: | [run_workflow_tests_run2_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537642) |
|:green_circle: | [run_workflow_tests_run3_mc](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537641) |
|:green_circle: | [run_workflow_tests_run3_data](https://gitlab.cern.ch/acts/acts-athena-ci/-/jobs/31537640) |

**Comment by paulgessinger** [comment]
Test failure was system-related.

**Pull request merged**